### ISSUE TITLE
sed, grep, awk

### ISSUE THREAD

**Issue by tjt263** [firasdib/Regex101#1230]
## Flavor Request

- [`sed`](https://www.gnu.org/software/sed/)
- [`grep`](https://www.gnu.org/software/grep/)
- [`awk`](https://www.gnu.org/software/gawk/)
- - -
No, they're not essential, and their syntaxes are similar to `php`, `perl`, `python`, etc.
Then again, nothing about Regex101 is strictly essential (i.e., it's not a C compiler or something of that nature); but that's not why we use it. We use it because we like to use it. It's nice, it's useful, it's convenient, and it makes our task easier. These would be welcome additions. Cheers.

**Comment by Doqnach** [comment]
See also #196 #165 #272

**Comment by zar3bski** [comment]
+1 It would be awsome to be able to test those with this tool. Noobs like me can literally spend hours dealing with `sed` based substitutions, specific required escaping, etc etc

**Comment by guest271314** [comment]
Found this trying to convert a JavaScript `RegExp` to `grep` pattern. This would be useful for the ability to view JavaScript `RegExp` in other language formats, if that is possible.

**Comment by Doqnach** [comment]
@guest271314 The code generator does little more than placing a written regex into a code template. It does no conversions for you. It is not guaranteed that the written regex will be correct for the specific language you generate the code for. The code generator is mostly there to give a starting point, where it is expected that the regex you wrote is correct for the chosen language.

**Comment by xenoterracide** [comment]
+1 to this please, I'm having a hell of a time debugging a `sed --regexp-extended` and the same pattern is fine in the provided languages.

**Comment by Ouims** [comment]
I am currently writing an ERE regex engine, which will be convertible to js easily, the 'only' thing left to do are alternatives and groups, I hope to have time and motivation to work on it more and convert it so it can be used on the website.
Alternatives are not too hard to do, but groups requires a bit more work.

**Comment by flatroof** [comment]
+1 awk [macox awk version 20070501 at present]

**Comment by Stephan972** [comment]
+1 sed
+1 grep

A great addition here...

**Comment by Codex-** [comment]
Would be extremely helpful to have these added :)

**Comment by working-name** [comment]
@Codex- can you give me an example of a regex where you need to count backslashes? I use grep and sed and I don't particularly remember having to fight the utilities, but I do have to fight bash's escaping needs.

**Comment by Codex-** [comment]
Not for counting backslashes specifically but more for validating the tokens are supported, i.e. the standard implementation of grep doesn't support `\d` (or perlre in general)

**Comment by pfdint** [comment]
In addition to #165, #196, #272, #404
Currently we go to https://sed.js.org/ which has the computation but not the rest of the resources that r101 provides. This appears to be the only calculator on the internet that implements regexes as described in https://www.gnu.org/software/sed/manual/sed.html

**Comment by tgross35** [comment]
I can probably add some context here (in addition to an upvote for the feature request). [From the grep man page:](https://linuxcommand.org/lc3_man_pages/grep1.html)

> grep  understands  three  different versions of regular expression syntax: “basic” (BRE), “extended” (ERE) and “perl” (PCRE).  In GNU grep  there  is  no  difference in available functionality between basic and extended  syntaxes.   In  other  implementations,  basic regular  expressions are less powerful.  The following description applies to extended regular  expressions;  differences  for  basic regular  expressions  are  summarized afterwards.  Perl-compatible regular  expressions  give  additional  functionality,   and   are documented  in  pcresyntax(3) and pcrepattern(3), but work only if PCRE is available in the system.

The rest of that page goes on to document BRE vs. ERE and is a great reference for anybody looking. Long story short, the only difference between the two is escaping and handling of special characters (?, +, (), {}, |). Support is as follows:

- awk: only supports ERE
- sed: supports BRE (default) or ERE (with -E)
- grep: supports BRE (default) or ERE (with -E). May support PCRE (with -P) but this is pretty uncommon; bourne shell doesn't support it

BRE (default) and ERE (activated with the -E flag) are the two implementations that are always available, and are also used for sed. -P for PCRE is out there but pretty uncommon, and not available with sed or awk anyway.

So, long story short, adding BRE and ERE support to the best string-related resource out there would be amazing!

**Comment by cheako** [comment]
Close this as a duplicate of: #165

**Comment by kj** [comment]
Not a duplicate. As explained above, that's for extended (`sed -E`), not basic (plain `sed`):

https://en.wikibooks.org/wiki/Regular_Expressions/POSIX_Basic_Regular_Expressions

**Comment by cheako** [comment]
I don't think having two separate issues adds anything to the conversation, one should be sufficient.

**Comment by kj** [comment]
I agree, but if one is closed, then the other should be clear that it's about both BRE and ERE.

**Comment by UrsineRaven** [comment]
To add a little to what @tgross35 said above,
Supporting `sed`, `grep`, `awk`, etc directly will involve handling different program-specific regular expression rules and may differ depending on implementation (e.g. GNU/Linux vs Mac OS). This seems very useful, but also potentially hard to maintain and/or confusing. So just supporting POSIX Basic and Extended regular expressions would probably be the best first step.
(relevant excerpts from man pages on a random Pi I was already connected to)
```
       s/regexp/replacement/
              Attempt to match regexp against the pattern space.   If  successful,  replace  that
              portion  matched with replacement.  The replacement may contain the special charac‐
              ter & to refer to that portion of the pattern space which matched, and the  special
              escapes \1 through \9 to refer to the corresponding matching sub-expressions in the
              regexp.

 Manual page sed(1) line 157 (press h for help or q to quit)
```
```
REGULAR EXPRESSIONS
       POSIX.2 BREs should be supported, but they aren't completely because of performance  prob‐
       lems.   The  \n  sequence in a regular expression matches the newline character, and simi‐
       larly for \a, \t, and other sequences.  The -E option switches to using  extended  regular
       expressions  instead;  it  has been supported for years by GNU sed, and is now included in
       POSIX.

 Manual page sed(1) line 236 (press h for help or q to quit)
```
```
REGULAR EXPRESSIONS
       [...]  In GNU grep  there  is  no  difference  in  available
       functionality  between  basic  and  extended  syntaxes.   In  other implementations, basic
       regular expressions are less powerful.  [...]

 Manual page grep(1) line 280 (press h for help or q to quit)
```

**Comment by kross-italk** [comment]
+1 for bash, vim dialects

**Comment by cheako** [comment]
> +1 for bash, vim dialects

I did recommend to close this issue in favor of the many others on the topic...  since that hasn't been done I can only assume you are supposed to open separate issues for both bash and vim.

**Comment by firasdib** [comment]
I have begun work on supporting Posix BRE, which should add support for grep, sed and awk. I am going to add code generators as well, so that you can use regex101 to create your BRE/ERE patterns, and use `grep` or `grep -E` without hassle.

**Comment by cheako** [comment]
I've been writing a test harness around whatever rust's libc crate loads, on my system that should be gnulibc.  The goal is to be ably the distill out any UB users might depend on.  The links can be run online.

* [Initial prototype, MVP](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=b4bca6bd1d4ae654b89289a423c2e9ae)
* [Safe rust abstraction](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=643bae01f848e6f20a8a90792673808d)

Then I've asked a LLM to generate 100 regexs and 8 strings for each and now I'm currently working through hooking them all up.

**Comment by firasdib** [comment]
Posix BRE will land in the next release!

**Comment by cheako** [comment]
So I've got tentative results of the [undefined behavior testing](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2024&gist=83c8333387c296209b8779a57acc6289).  The most surprising bit was the unicode support, though I'm not sure how it's supposed to work.  The current `Posix ERE` feature flags the unicode ranges as invalid, so that seems to be a regression.

As suspected I am confused, "modern" and "basic" REs are different enough that I'll have to redesign this.

**Comment by tgross35** [comment]
@cheako I have no idea what that code is intended to show but I'm fairly certain it is not checking for "undefined behavior" in the way that term is commonly understood.

**Comment by cheako** [comment]
@tgross35 I think so too.  The idea was to try and verify the output of Regex101 against implementations found in the wild.  So it's a framework to run a good number of regex against whatever implementation rust's libc crate will link with.  The "undefined behavior" I'm going for is not crashes or lockups, but instead oddities around edge cases and to sus out conditions where the definition(docs) doesn't match what the implementations actually do.
